python/giotto_json_reader.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python3
+
+import json
+import argparse
+
+
+def cells_json(data):
+    '''
+    >>> data = {
+    ...     "cell_1": {
+    ...         "locations": [1632.02, -1305.7],
+    ...         "factors": {
+    ...             "pleiden_clus": [3],
+    ...             "kmeans": [8]
+    ...         },
+    ...         "mappings": {
+    ...             "tsne": [-11.0776, 6.0311],
+    ...             "umap": [-6.858, 15.7691]
+    ...         }
+    ...     }
+    ... }
+    >>> cells = cells_json(data)
+    >>> cells.keys()
+    dict_keys(['cell_1'])
+    >>> cells['cell_1'].keys()
+    dict_keys(['mappings', 'genes', 'xy', 'factors', 'poly'])
+
+    '''
+    cell_dict = {}
+    for cell in data.keys():
+        factors = data[cell]['factors']
+
+        mappings_dict = {
+            't-SNE': data[cell]['mappings']['tsne'],
+            'UMAP': data[cell]['mappings']['umap']
+        }
+        factors_dict = {
+            'pleiden_clus': 'Cluster {}'.format(factors['pleiden_clus'][0]),
+            'kmeans': 'Cluster {}'.format(factors['kmeans'][0])
+        }
+        cell_dict[cell] = {
+            'mappings': mappings_dict,
+            'genes': {},
+            'xy': data[cell]['locations'],
+            'factors': factors_dict,
+            'poly': []
+        }
+
+    return cell_dict
+
+
+def factors_json(data):
+    '''
+    >>> data = {
+    ...     "cell_1": {
+    ...         "locations": [1632.02, -1305.7],
+    ...         "factors": {
+    ...             "pleiden_clus": [3],
+    ...             "kmeans": [8]
+    ...         },
+    ...         "mappings": {
+    ...             "tsne": [-11.0776, 6.0311],
+    ...             "umap": [-6.858, 15.7691]
+    ...         }
+    ...     }
+    ... }
+    >>> factors = factors_json(data)
+    >>> factors.keys()
+    dict_keys(['pleiden_clus', 'kmeans'])
+    >>> factors['kmeans'].keys()
+    dict_keys(['map', 'cells'])
+
+    '''
+    pleiden_clusters_set = {
+        get_factor(cell, 'pleiden_clus') for cell in data.values()
+    }
+    kmeans_clusters_set = {
+        get_factor(cell, 'kmean_clus') for cell in data.values()
+    }
+
+    pleiden_clusters_list = list(pleiden_clusters_set)
+    kmeans_clusters_list = list(kmeans_clusters_set)
+
+    pleiden_cells = {}
+    kmeans_cells = {}
+
+    for cell_id in data:
+        factors = data[cell_id]['factors']
+        pleiden_cells[cell_id] = pleiden_clusters_list.index(
+            factors['pleiden_clus'][0]
+        )
+        kmeans_cells[cell_id] = kmeans_clusters_list.index(
+            factors['kmeans'][0]
+        )
+
+    factors_dict = {
+        'pleiden_clus': {
+            'map': ['Cluster {}'.format(c) for c in pleiden_clusters_list],
+            'cells': pleiden_cells
+        },
+        'kmeans': {
+            'map': ['Cluster {}'.format(c) for c in kmeans_clusters_list],
+            'cells': kmeans_cells
+        }
+    }
+
+    return factors_dict
+
+
+def get_factor(cell, factor_name):
```

Comment:
I think this is the thing that started out as a lambda? Remember that you can define functions inside other functions... sometime not a good idea, but here, since it's short, it might make things clearer.